numpy/lib/function_base.py

```diff
@@ -1320,6 +1320,477 @@ def gradient(f, *varargs, **kwargs):
+            "Shape of array too small to calculate a numerical gradient, "
+            "at least two elements are required.")
+        
+        # Numerical differentiation: 1st order edges, 2nd order interior
+        if y.shape[axis] == 2:# or edge_order == 1:
+            # Use first order differences for time data
+            out = np.empty_like(y, dtype=otype)
+            
+            slice1[axis] = slice(1, -1)
+            slice2[axis] = slice(2, None)
+            slice3[axis] = slice(None, -2)
+            # 1D equivalent -- out[1:-1] = (y[2:] - y[:-2])/2.0
+            out[slice1] = (y[slice2] - y[slice3])/2.0
+            
+            slice1[axis] = 0
+            slice2[axis] = 1
+            slice3[axis] = 0
+            # 1D equivalent -- out[0] = (y[1] - y[0])
+            out[slice1] = (y[slice2] - y[slice3])
+            
+            slice1[axis] = -1
+            slice2[axis] = -1
+            slice3[axis] = -2
+            # 1D equivalent -- out[-1] = (y[-1] - y[-2])
+            out[slice1] = (y[slice2] - y[slice3])
+            
+            # Numerical differentiation: 2st order edges, 2nd order interior
+        else:
+    	    # Use second order differences where possible
+            out = np.empty_like(y, dtype=otype)
+            
+            slice1[axis] = slice(1, -1)
+            slice2[axis] = slice(2, None)
+            slice3[axis] = slice(None, -2)
+            # 1D equivalent -- out[1:-1] = (y[2:] - y[:-2])/2.0
+            out[slice1] = (y[slice2] - y[slice3])/2.0
+            
+            slice1[axis] = 0
+            slice2[axis] = 0
+            slice3[axis] = 1
+            slice4[axis] = 2
+            # 1D equivalent -- out[0] = -(3*y[0] - 4*y[1] y[2]) / 2.0
+            out[slice1] = -(3.0*y[slice2] - 4.0*y[slice3] + y[slice4])/2.0
+            
+            slice1[axis] = -1
+            slice2[axis] = -1
+            slice3[axis] = -2
+            slice4[axis] = -3
+            # 1D equivalent -- out[-1] = (3*y[-1] - 4*y[-2] y[-3])
+            out[slice1] = (3.0*y[slice2] - 4.0*y[slice3] + y[slice4])/2.0
+        
+        # divide by step size
+        out /= dx[axis]
+        
+        if i == 0:
+            out0 = out
+        elif i == 1:
+            out1 = out
+        elif i == 2:
+            out2 = out
+        elif i == 3:
+            out3 = out
+        elif i == 4:
+            out4 = out
+        else:
+            out5 = out
+        
+        # reset the slice object in this dimension to ":"
+        slice1[axis] = slice(None)
+        slice2[axis] = slice(None)
+        slice3[axis] = slice(None)
+        slice4[axis] = slice(None)
+
+    
+    outvals[0] = out0 - out1
+    outvals[1] = out2 - out3
+    outvals[2] = out4 - out5
+    return outvals
+
+
+def laplace_s(f):
+    """
+    Return the laplacian of an N-dimensional array.
+
+    Computes the laplacian of an N-dimensional numpy array, as a
+    concatenation of the numpy functions gradient and divergence. 
```

Comment:
Instead of 'concatenation' do you mean 'composition' here? Or this could be jargon that is unfamiliar to me, for example the composition of functions might be written as the concatenation of the symbols representing the functions. And what is the meaning of the `_s` in `laplace_s`?